DOCSP-30524 Fundamentals > Connection Guide > Connection Options (#27)
mongodb/docs-rust: source/fundamentals/connections/connection-options.txt

.. _rust-connection-options:

> =====================
> Connection Options
> =====================

Reviewer: S: align with the length of the title
```suggestion
===============
Connection Options
===============
```

> .. default-domain:: mongodb

Reviewer: S: this is a deprecated functionality, remove

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> This section explains several common MongoDB connection and authentication
> options. You can pass the connection options as parameters of the connection
> URI to specify the behavior of the client.

Reviewer: S: Start any fundamentals page with an "In this guide..." statement

For example, this guide could start something like:
```suggestion
In this guide, you can learn about the connection options available in the {+driver-short+}.
You can set connection options as parameters of your connection URI to specify how your client behaves
while connected to MongoDB.
```

.. list-table::
   :header-rows: 1
   :widths: 34 10 12 44

   * - Option Name
     - Type
     - Default Value
     - Description

>    * - **connect_timeout**

Reviewer: I: I noticed that these connection option names are the ones from the Driver API. We want to actually include the connection string options on this page, which can be found [here](https://www.mongodb.com/docs/manual/reference/connection-string/#connection-string-options). So, to reiterate, we want only the options that are in the Rust Driver ClientOptions API -- but we are actually describing the ones in the connection string format

Reviewer: The list of options you should include on this page is given here with the mapping -- https://docs.rs/mongodb/2.6.0/src/mongodb/client/options/mod.rs.html#1017-1070

   * - **authMechanism**
     - string
     - None
     - Specifies which authentication mechanism to use.
       If empty, one will be negotiated with the server.

   * - **authMechanismProperties**
     - string
     - None
     - Specifies more properties for the given mechanism.

   * - **authSource**
     - string
     - See description.
     - The database used to authenticate.
       This applies to all authentication mechanisms and defaults to ``admin`` in SCRAM authentication mechanisms,
       ``$external`` for GSSAPI and MONGODB-X509, and the database name or ``$external`` for PLAIN.

   * - **compressors**
     - comma separated list of strings, for example, "snappy,zlib,zstd"
     - None
     - Specifies compressors that the ``Client`` is willing to use in the order they are specified in the configuration.
       The ``Client`` sends this list of compressors to the server.
       The server responds with the intersection of its supported list of compressors.
       The order of compressors indicates preference of compressors.

   * - ***connectTimeoutMS***
     - non-negative integer

>      - 10 seconds

Reviewer: S: since this is in MS, show that the default value is actually 10000 (10 seconds). Applies to other MS options, too

>      - Specifies the connection timeout passed to each underlying TCP stream when attempting to connect to the server.

Reviewer: ```suggestion
     - Specifies the connection timeout, in milliseconds, passed to each underlying TCP stream when attempting to connect to the server.
```

   * - **direct**
     - boolean
     - ``false``

>      - Specifies whether the ``Client`` will directly connect to a single host rather than automatically discovering all servers in the cluster.
> 

Reviewer: ```suggestion
     - Specifies whether the ``Client`` instance directly connects to a single host instead of discovering and connecting to all of the servers in the cluster.

```

   * - **heartbeatFrequencyMS**
     - integer greater than or equal to 500
     - 10 seconds

>      - Specifies the each monitoring thread waits between performing server checks.

Reviewer: ```suggestion
     - Specifies the amount of time, in milliseconds, that each monitoring thread waits between performing server checks.
```

   * - **journal**
     - boolean
     - ``false``

>      - Requests acknowledgment that the operation has propagated to the on-disk journal.

Reviewer: ```suggestion
     - Requests acknowledgment that the operation propagated to the on-disk journal.
```

   * - **localThresholdMS**
     - non-negative integer
     - 15 milliseconds

>      - Specifies how much longer the average round trip time between the driver and server is allowed to be
>        compared to the least round trip time of all the suitable servers.
>        A value of zero indicates that there is no latency window, so only the server with the lowest average round trip time is eligible.

Reviewer: ```suggestion
     - Specifies how much longer, in milliseconds, that the average round trip time between the driver and server can be
       compared to the least round trip time of all the suitable servers.
       A value of ``0`` indicates that there is no latency window, so only the server with the lowest average round trip time is eligible.
```

   * - **maxIdleTimeMS**
     - non-negative integer
     - 0

>      - Specifies the time that a connection can remain idle in a connection pool before being closed.
>        A value of zero indicates that connections are not closed due to being idle.

Reviewer: ```suggestion
     - Specifies the amount of time, in milliseconds, that a connection can remain idle in a connection pool before being closed.
       A value of ``0`` indicates that connections are not closed due to being idle.
```

   * - **maxStalenessSeconds**

>      - -1, or an integer greater than or equal 90

Reviewer: ```suggestion
     - -1, or any integer greater than or equal to ``90``
```

     - ``-1``

>      - Specifies the maximum lag behind the primary that a secondary can have in order to be considered for the given operation.
>        Any secondaries lagging behind more than this will not be considered for the operation.
> 
>        Must be at least 90 seconds; if less, the operation will return an error.
>        ``-1`` means there is no maximum.

Reviewer: ```suggestion
     - Specifies the maximum lag, in seconds, behind the primary node that a secondary node can be to be 
       considered for the given operation.

       The value of this option must be at least ``90``, or the operation raises an error.
       A value of ``-1`` means there is no maximum lag.
```

   * - **maxPoolSize**
     - non-negative integer
     - 10

>      - Specifies the maximum number of connections that the ``Client`` is allowed to create in a connection pool for a given server.

Reviewer: S: add "instance" in all places where needed
```suggestion
     - Specifies the maximum number of connections that the ``Client`` instance is allowed to create in a connection pool for a given server.
```

>      - Specifies the maximum number of connections that the ``Client`` is allowed to create in a connection pool for a given server.
>        If an operation is attempted on a server while **maxPoolSize** connections are checked out,
>        the operation will block until an in-progress operation finishes and its connection returns to the pool.

Reviewer: ```suggestion
     - Specifies the maximum number of connections that the ``Client`` can create in a connection pool for a given server.
       If you attempt an operation while the value of ``maxPoolSize`` connections are checked out,
       the operation waits until an in-progress operation finishes and the connection returns to the pool.
```

   * - **minPoolSize**
     - non-negative integer
     - 0

>      - Specifies the minimum number of connections that will be available in a server's connection pool at a given time.
>        If fewer than **minPoolSize** connections are in the pool, connections will be added to the pool in the background
>        until **minPoolSize** is reached.

Reviewer: ```suggestion
     - Specifies the minimum number of connections that are available in a server's connection pool at a given time.
       If fewer than ``minPoolSize`` connections are in the pool, the server adds connections background up to the value of ``minPoolSize``.
```

   * - **readConcernLevel**
     - string
     - None

>      - Specifies the default read concern for operations performed on the ``Client``.
>        See the `ReadConcern <https://docs.rs/mongodb/latest/mongodb/options/struct.ReadConcern.html>`__
>        type documentation for more details.

Reviewer: ```suggestion
     - Specifies the default read concern for operations performed on the ``Client``.
       To learn more, see the `ReadConcern <https://docs.rs/mongodb/latest/mongodb/options/struct.ReadConcern.html>`__
       type API documentation.
```

   * - **readPreferenceField**
     - string
     - ``primary``

>      - Specifies how the driver routes a read operation to members of a replica set.
>        For more information see the :manual:`Read Preference documention </core/read-preference>`.

Reviewer: S: change all server manual links to this format

```suggestion
     - Specifies how the driver routes a read operation to members of a replica set.
       To learn more, see :manual:`Read Preference </core/read-preference>` in the Server manual.
```

   * - **readPreferenceTags**

>      - comma-separated key:value pairs, for example, "dc:ny,rack:1"

Reviewer: ```suggestion
     - A list of comma-separated key-value pairs
```

     - None
     - Specifies which replica set members are considered for operations.
       Each instance of this key is a separate tag set.

>        Each tag set will be checked in order until one or more servers is found with each tag in the set.

Reviewer: ```suggestion
      The driver checks each tag set until it finds one or more servers with each tag in the set.
```

   * - **replicaSet**
     - string
     - None
     - Specifies the name of the replica set that the ``Client`` connects to.

   * - **retryReads**
     - boolean
     - ``true``

>      - Specifies whether the client will retry a read operation if the operation fails.

Reviewer: S: use present tense (applies throughout)
```suggestion
     - Specifies whether the client retries a read operation if the operation fails.
```

   * - **serverSelectionTimeoutMS**
     - non-negative integer

>      - 30 seconds

Reviewer: S: give in MS

>      - Specifies the time the ``Client`` waits when attempting to select a server for an operation before timing out.

Reviewer: ```suggestion
     - Specifies the amount of time, in milliseconds, that the ``Client`` instance waits when attempting to select a server for an operation before timing out.
```

   * - **tls**
     - boolean
     - ``false``

>      - Specifies the TLS configuration for the ``Client`` to use in its connections with the server.
>        By default, TLS is off.

Reviewer: ```suggestion
     - Specifies the TLS configuration for the ``Client`` to use in its connections with the server.
       By default, TLS is disabled.
```

Author: Vale did not like this.

>    * - **tlsAllowInvalidCertificates** (or **tlsInsecure**)

Reviewer: S: separate these options into their own rows

     - boolean
     - ``true``
     - Specifies whether the ``Client`` returns an error if the server presents an invalid certificate.
       (Do not set to ``true`` in production.
       This is only for testing.)

   * - **tlsCAFile**
     - string
     - See description.
     - Specifies the path to the CA file that the ``Client`` uses for TLS.

>        If none is specified, then the driver will use the Mozilla root certificates from the ``webpki-roots`` crate.

Reviewer: ```suggestion
       If you do not specify this option, the driver uses the Mozilla root certificates from the ``webpki-roots`` crate.
```

   * - **tlsCertificateKeyFile**
     - string
     - None
     - Specifies the path to the certificate file that the ``Client`` presents to the server to verify its identify.

>        If none is specified, then the ``Client`` will not attempt to verify its identity to the server.

Reviewer: ```suggestion
       If you do not set this option, the ``Client`` instance does not attempt to verify its identity to the server.
```

   * - **w**
     - non-negative integer or string
     - None

>      - Requests acknowledgment that the operation has propagated to a specific number or variety of servers.

Reviewer: S:Link to write concern documentation

.. _rust-connection-options:

==================
Connection Options
==================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

In this guide, you can learn about MongoDB connection and authentication
options. You can pass connection options as parameters of the connection
string URI to specify the behavior of the client. To learn more about connection strings, see

> :manual:`Connection String URI Format </reference/connection-string>` in the Server

Reviewer: S: Update the page name (I think they recently changed it)

```suggestion
:manual:`Connection Strings </reference/connection-string>` in the Server manual.
```

.. TODO link to the authentication mechanisms guide in authMechanism below

.. list-table::
   :header-rows: 1
   :widths: 30 12 12 46

   * - Option Name
     - Accepted Values
     - Default Value
     - Description

   * - **appName**
     - Any string
     - None
     - Specifies the application name that the ``Client`` instance sends to the server as part of the handshake.
       Specifying an ``appName`` can help you use the server logs to determine which ``Client`` instance is connected
       to the server.

   * - **authMechanism**
     - Any string
     - None
     - Specifies which authentication mechanism to use.

>        If empty, one will be negotiated with the server.

Reviewer: ```suggestion
       If you do not specify this option, the driver uses the default authentication mechanism.
```

   * - **authMechanismProperties**
     - Any string
     - None

>      - Specifies additional properties for the authentication mechanism

Reviewer: S: make sure all descriptions have correct punctuation
```suggestion
     - Specifies additional properties for the authentication mechanism set in the ``authMechanism`` option.
```

     - Specifies additional properties for the authentication mechanism set in the ``authMechanism`` option.
       specified in the ``authMechanism`` option.

   * - **authSource**
     - String

>      - See description.

Reviewer: S: remove periods from all entries in this column

       This option defaults to ``admin`` for SCRAM-based authentication mechanisms,
       ``$external`` for the ``MONGODB-X509`` mechanism, and the database name or
       ``$external`` for the ``PLAIN`` mechanism.

   * - **compressors**
     - A comma-separated list of strings
     - None

>      - Specifies compressors that the ``Client`` instance uses in the specified order.

Reviewer: S: link to the network compression fundamentals page in the description

   * - ***connectTimeoutMS***
     - Non-negative integer
     - 10000
     - Specifies the connection timeout, in milliseconds, passed to each underlying TCP stream when attempting to connect to the server.

   * - **direct**
     - Boolean
     - ``false``
     - Specifies whether the ``Client`` instance directly connects to a single host
       instead of discovering and connecting to all of the servers in the cluster.

   * - **heartbeatFrequencyMS**
     - Integer greater than or equal to 500
     - 10000
     - Specifies the amount of time, in milliseconds, that each monitoring thread waits between performing server checks.

   * - **journal**
     - Boolean
     - ``false``
     - Requests acknowledgment that the operation propagated to the on-disk journal.

   * - **localThresholdMS**
     - Non-negative integer
     - 15

>      - Specifies how much longer, in milliseconds, that the average round trip time between the driver and server can be

Reviewer: ```suggestion
     - Specifies how much longer, in milliseconds, that the average round trip time between the driver and server can take
```

>        compared to the least round trip time of all the suitable servers.

Reviewer: ```suggestion
       compared to the shortest round trip time of all the suitable servers.
```

       A value of ``0`` indicates that there is no latency window, so only the server with the lowest average round trip time is eligible.

   * - **maxIdleTimeMS**
     - Non-negative integer
     - 0

>      - Specifies the amount of time, in milliseconds, that a connection can remain idle in a connection pool before being closed.

Reviewer: S; avoid passive voice
```suggestion
     - Specifies the amount of time, in milliseconds, that a connection can remain idle in a connection pool the server closes it.
```

       considered for the given operation.

       The value of this option must be at least ``90``, or the operation raises an error.
       A value of ``-1`` means there is no maximum lag.

   * - **maxPoolSize**
     - Non-negative integer
     - 10
     - Specifies the maximum number of connections that the ``Client`` instance can create in a connection pool for a given server.
       If you attempt an operation while the value of ``maxPoolSize`` connections are checked out,
       the operation waits until an in-progress operation finishes and the connection returns to the pool.

   * - **minPoolSize**
     - Non-negative integer
     - 0
     - Specifies the minimum number of connections that are available in a server's connection pool at a given time.
       If fewer than ``minPoolSize`` connections are in the pool,
       the server adds connections in the background up to the value of ``minPoolSize``.

   * - **readConcernLevel**
     - Any string
     - None
     - Specifies the default read concern for operations performed on the ``Client`` instance.
       To learn more, see the `ReadConcern <https://docs.rs/mongodb/latest/mongodb/options/struct.ReadConcern.html>`__

>        type API documentation.

Reviewer: S: Link to the server manual instead
"To learn more, see :manual:`Read Concern </reference/read-concern/>` in the Server manual.

>    * - **readPreferenceField**

Reviewer: I: I think there is a typo in the rust driver docs, this option is actually called `readPreference` https://www.mongodb.com/docs/manual/reference/connection-string/#read-preference-options

Also, the values cannot be any string - see the table here: https://www.mongodb.com/docs/manual/reference/connection-string/#read-preference-options:~:text=Connection%20Option-,Description,-readPreference

please check for all options that the accepted values are correct

     - A list of comma-separated key-value pairs
     - None
     - Specifies which replica set members are considered for operations.
       Each instance of this key is a separate tag set.
       The driver checks each tag set until it finds one or more servers with each tag in the set.

   * - **replicaSet**
     - Any string
     - None
     - Specifies the name of the replica set that the ``Client`` instance connects to.

   * - **retryReads**
     - Boolean
     - ``true``
     - Specifies whether the client retries a read operation if the operation fails.

   * - **serverSelectionTimeoutMS**
     - Non-negative integer
     - 30000
     - Specifies the amount of time, in milliseconds, that the ``Client`` instance waits when attempting to select a server for an operation before timing out.

   * - **tls**
     - Boolean
     - ``false``
     - Specifies the TLS configuration for the ``Client`` instance to use in its connections with the server.
       By default, TLS is off.

   * - **tlsAllowInvalidCertificates**
     - Boolean
     - ``true``
     - Specifies whether the ``Client`` instance returns an error if the server presents an invalid certificate.

>        (Do not set to ``true`` in production.
>        This is only for testing.)

Reviewer: ```suggestion
       We recommend that you set this option to ``true`` only in testing environments, to avoid creating vulnerabilities in your application.
```

>      - ``true``

Reviewer 2: `tlsAllowInvalidCertificates` is false by default.

Replace ``true`` with ``false``.

   * - **tlsCAFile**
     - Any string
     - See description.

>      - Specifies the path to the CA file that the ``Client`` instance uses for TLS.

Reviewer: ```suggestion
     - Specifies the path to the certificate authority (CA) file that the ``Client`` instance uses for TLS.
```

       If you do not specify this option, the driver uses the Mozilla root certificates from the ``webpki-roots`` crate.

   * - **tlsCertificateKeyFile**
     - Any string
     - None
     - Specifies the path to the certificate file that the ``Client`` instance presents to the server to verify its identify.
       If you do not set this option, the ``Client`` instance does not attempt to verify its identity to the server.

   * - **tlsInsecure**
     - Boolean
     - ``true``
     - Specifies whether the ``Client`` instance returns an error if the server presents an invalid certificate.

>        (Do not set to ``true`` in production.
>        This is only for testing.)

Reviewer: ```suggestion
       We recommend that you set this option to ``true`` only in testing environments, to avoid creating vulnerabilities in your application.
```

>      - ``true``

Reviewer 2: `tlsInsecure` is false by default.

Replace ``true`` with ``false``.

   * - **w**
     - Non-negative integer or string
     - None
     - Requests acknowledgment that the operation has propagated to a specific number or variety of servers.

>        To learn more, see :manual:`Write Concern </core/read-preference>` in the Server manual.

Reviewer: I: fix the link to go to the right place -- please double check all links

   * - **wTimeoutMS**
     - Non-negative integer
     - No timeout

>      - Specifies a time limit for the write concern.
>        If an operation has not propagated to the requested level within the time limit, an error will return.

Reviewer: ```suggestion
     - Specifies a time limit, in milliseconds, for the write concern.
       If an operation has not propagated to the requested level within the time limit, the driver raises an error.
```

   * - **zlibCompressionLevel**
     - Integer between -1 and 9 (inclusive)
     - ``-1``
     - Specifies the level field of the ``zlib`` compression when in use.

>        See the `compressor API documentation <{+api+}/options/enum.Compressor.html#variant.Zlib>`__ for more information.
>        (Requires enabling the ``zlib-compression`` feature.)
>        ``-1`` means no timeout.

Reviewer: S: link to the network compression page for Rust that is merged. Also, double check the meaning of the -1 value -- I thought these levels referred to the level of compression, not a timeout

```suggestion
       To learn more about network compression, see the :ref:`rust-network-compression` guide.
```

       specified in the ``authMechanism`` option.

   * - **authSource**
     - String
     - See description
     - Specifies the database used to authenticate.
       This option defaults to ``admin`` for SCRAM-based authentication mechanisms,
       ``$external`` for the ``MONGODB-X509`` mechanism, and the database name or
       ``$external`` for the ``PLAIN`` mechanism.

   * - **compressors**
     - A comma-separated list of strings
     - None
     - Specifies compressors that the ``Client`` instance uses in the specified order.
       To learn more about network compression, see the :ref:`rust-network-compression` guide.

   * - ***connectTimeoutMS***
     - Non-negative integer
     - 10000
     - Specifies the connection timeout, in milliseconds, passed to each underlying TCP stream when attempting to connect to the server.

   * - **direct**
     - Boolean
     - ``false``
     - Specifies whether the ``Client`` instance directly connects to a single host
       instead of discovering and connecting to all of the servers in the cluster.

   * - **heartbeatFrequencyMS**
     - Integer greater than or equal to 500

>      - 10000

Reviewer: S: I think we should decide on a convention for formatting numbers on this page, and I am leaning towards monospacing all numbers that actually refer to possible values for an option. For example, if you described one of the Millisecond options, you could monospace `10000` but not monospace 10 seconds. 
i.e. 
Default value: ``100000`` (10 seconds)

Reviewer: Applies throughout. Would be helpful for each Millisecond option to provide the conversion in seconds

     - A list of comma-separated key-value pairs
     - None
     - Specifies which replica set members are considered for operations.
       Each instance of this key is a separate tag set.
       The driver checks each tag set until it finds one or more servers with each tag in the set.

   * - **replicaSet**
     - String
     - None
     - Specifies the name of the replica set that the ``Client`` instance connects to.

   * - **retryReads**
     - Boolean
     - ``true``
     - Specifies whether the client retries a read operation if the operation fails.

   * - **serverSelectionTimeoutMS**
     - Non-negative integer
     - 30000
     - Specifies the amount of time, in milliseconds, that the ``Client`` instance waits when attempting to select a server for an operation before timing out.

   * - **tls**
     - Boolean
     - ``false``
     - Specifies the TLS configuration for the ``Client`` instance to use in its connections with the server.
       By default, TLS is off.

   * - **tlsAllowInvalidCertificates**
     - Boolean
     - ``true``
     - Specifies whether the ``Client`` instance returns an error if the server presents an invalid certificate.

>        We recommend that you set this option to ``true`` only in testing environments, to avoid creating vulnerabilities in your application.

Reviewer: ```suggestion
       We recommend that you set this option to ``true`` only in testing environments to avoid creating vulnerabilities in your application.
```

   * - **tlsCAFile**
     - String
     - See description.
     - Specifies the path to the certificate authority (CA) file that the ``Client`` instance uses for TLS.
       If you do not specify this option, the driver uses the Mozilla root certificates from the ``webpki-roots`` crate.

   * - **tlsCertificateKeyFile**
     - String
     - None
     - Specifies the path to the certificate file that the ``Client`` instance presents to the server to verify its identify.
       If you do not set this option, the ``Client`` instance does not attempt to verify its identity to the server.

   * - **tlsInsecure**
     - Boolean
     - ``true``
     - Specifies whether the ``Client`` instance returns an error if the server presents an invalid certificate.

>        We recommend that you set this option to ``true`` only in testing environments, to avoid creating vulnerabilities in your application.

Reviewer: ```suggestion
       We recommend that you set this option to ``true`` only in testing environments to avoid creating vulnerabilities in your application.
```

   * - **w**
     - Non-negative integer or string
     - None
     - Requests acknowledgment that the operation has propagated to a specific number or variety of servers.
       To learn more, see :manual:`Write Concern </reference/write-concern>` in the Server manual.

   * - **wTimeoutMS**
     - Non-negative integer
     - No timeout
     - Specifies a time limit, in milliseconds, for the write concern.
       If an operation has not propagated to the requested level within the time limit, the driver raises an error.

   * - **zlibCompressionLevel**
     - Integer between -1 and 9 (inclusive)
     - ``-1``

>      - Specifies the level field of the ``zlib`` compression when in use.
>        See the `compressor API documentation <{+api+}/options/enum.Compressor.html#variant.Zlib>`__ for more information.
>        (Requires enabling the ``zlib-compression`` feature.)
>        ``-1`` sets the default compression level (usually 6).
>        ``0`` means no compression.
>        To learn more about network compression, see the :ref:`rust-network-compression` guide.

Reviewer: ```suggestion
     - Specifies the level field of the ``zlib`` compression if you use that compressor.
       Setting a value of``-1`` selects the default compression level (``6``).
       Setting a value of ``0`` specifies no compression, and setting a value of ``9`` specifies maximum
       compression.
       To learn more about network compression, see the :ref:`rust-network-compression` guide.
```
